### ISSUE TITLE
Range of EOF incorrect

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#2169]
The `Range` of the EOF token sit right on top of the last character of the previous token. I would say it sits one position to the right of it. JavaCC bug?

**Comment by MysterAitch** [comment]
I think I've been seeing this (or something similar) recently -- specifically, the end position of the `CompilationUnit` range seems to be wrong. 

Does this sound like what you report, or is this something different?


What I observe seems to be newline related, where if the last characters of the file are a newline they are treated as characters on the second-to-last line, as opposed to triggering a newline. This is only true where the last line is empty.

**Example on windows (which uses crlf):**

CU range: (1,1) to (1,17) -- no empty line
```java 
public class X {}
```

CU range: (1,1) to (1,19) -- single empty line
```java
public class X {}

```

CU range: (1,1) to (2,2) -- two empty lines
```java
public class X {}


```

**Comment by matozoid** [comment]
Nah, this issue is specifically about the EOF token, but maybe these are connected?

**Comment by matozoid** [comment]
It would be nice to write a test case PR for your cases, they get messed up with the formatting here.

**Comment by MysterAitch** [comment]
Just pushed the following testcase to #2648

Apologies if this is a silly question, but is there a reason that the EOF token is excluded from the compilation unit's token range? 

Presumably the EOF token would exist on the line following the final linebreak, but if that is skipped somewhere then the last character visible does exist at the end of the second-to-last line (noting that positions thus ranges refer only to characters, not the spaces _before/between/after_ characters).

If it does look to you that this is a separate and not linked, please do let me know and I'll create a separate issue for it.


```java
    private static Stream<Arguments> cuRangeProvider() {
        return Stream.of(
                Arguments.of("class X{}", new Position(1, 1), new Position(1, 9)),     // PASSES
                Arguments.of("class X{}\n", new Position(1, 1), new Position(2, 1)),   // FAILS -- actual: (line 1,col 10)
                Arguments.of("class X{}\n\n", new Position(1, 1), new Position(3, 1))  // FAILS -- actual: (line 2,col 1)
        );
    }

    @ParameterizedTest
    @MethodSource("cuRangeProvider")
    void test(String x, Position expectedRangeBegin, Position expectedRangeEnd) {
        final CompilationUnit cu = parse(x);
        final Range range = cu.getRange().get();

        assertEquals(expectedRangeBegin, range.begin);
        assertEquals(expectedRangeEnd, range.end);
    }
```

**Comment by matozoid** [comment]
You could extend that test with a check of the range of the last token in the CU's tokenrange (which ought to be EOL?)

Positions have been rather buggy in the past, and it wouldn't surprise me if there are still bugs. If EOL should be included or not, well, I guess? Otherwise it belongs to no Node. If I look at the grammar the EOL should be included already: https://github.com/javaparser/javaparser/blob/master/javaparser-core/src/main/javacc/java.jj#L657

so this will need a debugger.
